Read+Write sides of the replay determinism:

ifesdjeen opened the pull request:
* Persist intermediate Command inthe journal
      * Simplify AccordJournal by removing Framing
      * Save command outcomes to the log
      * Reconstruct latest command state from the log entries
      * Replace `SerializerSupport#reconstruct` with log reconstruction

maedhroz committed:
CEP-15: Multi-Partition Transaction CQL Support (Alpha)

patch by Caleb Rackliffe; reviewed by David Capwell for CASSANDRA-17719

Co-authored-by: Caleb Rackliffe <calebrackliffe@gmail.com>
Co-authored-by: Blake Eggleston <bdeggleston@gmail.com>
Co-authored-by: David Capwell <dcapwell@apache.org>

belliottsmith committed:
Partial Replication of Transaction State

The replication of transactions and their processing state must be optimal, i.e. linearly proportional to the size of the transaction and the number of keys involved. This patch ensures a transaction is replicated in slices, with each slice being written (replication factor) times only to the shard that owns the keys, and only once (except in the case the transaction is recovered).

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-17109

bdeggleston committed:
CEP-15 (C*): Messaging and storage engine integration

patch by Blake Eggleston; reviewed by Benedict Elliott Smith, David Capwell for CASSANDRA-17103

belliottsmith committed:
CEP-15: Routables
  - Integrate accord-core changes for CASSANDRA-18087

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-18087

dcapwell committed:
Command table now uses a local version added to the payload rather than rely on a version column, and fixed a few feedback related issues

patch by David Capwell; reviewed by Blake Eggleston for CASSANDRA-17103

belliottsmith committed:
Transaction Expiration

Ensure old or abandoned transactions are invalidated wherever possible.

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-18041

belliottsmith committed:
Invalidation fixes/improvements
 - Integrate accord-core changes for CASSANDRA-18057

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-18057

belliottsmith committed:
Refactor response tracking to improve efficiency and clarity; introduce dedicated property tests; re-activate fast-path during range movements

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-18056

aweisberg committed:
Support CAS and serial read on Accord

patch by Ariel Weisberg; Reviewed by Caleb Rackliffe for CASSANDRA-18100

belliottsmith committed:
Shard local CommandStores on contiguous ranges

patch by Benedict; reviewed by Aleksey Yeschenko for CASSANDRA-18142

belliottsmith committed:
Refactor Timestamp/TxnId
 - Combine real and logical into a single 64-but HLC
 - Introduce 16 flag bits
 - Pack epoch (48-bits), HLC (64-bits) and flags (16-bits) into two longs in memory

patch by Benedict; reviewed by Aleksey Yeschenko for CASSANDRA-18172

belliottsmith committed:
Introduce RangeDeps

Refactor Deps into KeyDeps and RangeDeps

patch by Benedict; reviewed by Aleksey Yeschenko for CASSANDRA-18173

belliottsmith committed:
Introduce Range transactions

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-18174

belliottsmith committed:
ninja: pick a cassandra-accord SHA to build against to decouple source trees

dcapwell committed:
CASSANDRA-18154: CEP-15: Enhance returning SELECT to allow partition and clustering IN clauses to return multiple partitions/rows

dcapwell committed:
CEP-15: (C*) Enhance returning SELECT to allow partition and clustering IN clauses to return multiple partitions/rows

patch by David Capwell; reviewed by Caleb Rackliffe for CASSANDRA-18154

aweisberg committed:
Accord support for live migration

Patch by Ariel Weisberg; Reviewed by Blake Eggleston for CASSANDRA-18129

maedhroz committed:
Post-rebase fixes around CASSANDRA-18099 and CASSANDRA-18114 (among other things)

bdeggleston committed:
add AsyncChain implementations and tests

Patch by Blake Eggleston; Reviewed by David Capwell & Benedict Elliott Smith for CASSANDRA-18004

maedhroz committed:
Feature Flag for Accord Transactions

patch by Caleb Rackliffe; reviewed by David Capwell for CASSANDRA-18195

iamaleksey committed:
Switch Node.Id from long to int

patch by Aleksey Yeschenko; reviewed by Benedict Elliott Smith for
CASSANDRA-18135

dcapwell committed:
Ninja: CASSANDRA-18214 added rat checks to Accord, but our metadata files .BRANCH and .REPO do not have a license, causing the build to fail; exclude rat when building accord

patch by David Capwell; reviewed by Ariel Weisberg for CASSANDRA-18214

maedhroz committed:
use equals() rather than reference equality for excluding ourselves from the dependency builder

patch by Benedict; reviewed by Caleb Rackliffe for CASSANDRA-18218

bdeggleston committed:
CEP-15/Accord Immutable State Refactor

Patch by Blake Eggleston; Reviewed by David Capwell and Benedict Elliott Smith for Cassandra-18192

jacek-lewandowski committed:
Fix null value handling for static columns

patch by <jacek-lewandowski>; reviewed by <maedhroz> and <dcapwell> for CASSANDRA-18241

iamaleksey committed:
CEP-15: Store PreAccept, Accept, Commit, and Apply messages
in a durable log before processing by CommandStores

patch by Aleksey Yeschenko; reviewed by David Capwell for
CASSANDRA-18344

jacek-lewandowski committed:
Fix statement validation against partition range queries

patch by Jacek Lewandowski; reviewed by Caleb Rackliffe for CASSANDRA-18240

dcapwell committed:
CEP-15: (C*) Add git submodule for Accord

patch by David Capwell; reviewed by Caleb Rackliffe, Michael Semb Wever for CASSANDRA-18204

dcapwell committed:
Switched away from https endpoints for git submodules in favor of relative paths during development

patch by David Capwell; reviewed by Caleb Rackliffe, Michael Semb Wever for CASSANDRA-18204

dcapwell committed:
Add support for prepared statements for accord transactions

patch by David Capwell; reviewed by Alex Petrov, Caleb Rackliffe, Jacek Lewandowski for CASSANDRA-18299

dcapwell committed:
Remove git hook for pre-push as it is redundant and causes issues when merging to mainline

patch by David Capwell; reviewed by Caleb Rackliffe for CASSANDRA-18309

dcapwell committed:
CEP-15: (Accord) Migrate Accord away from JDK random to a new interface RandomSource

patch by David Capwell; reviewed by Blake Eggleston for CASSANDRA-18213

jacek-lewandowski committed:
Improve transaction statement validation

patch by Jacek Lewandowski; reviewed by David Capwell and Caleb Rackliffe for CASSANDRA-18302

maedhroz committed:
fixing remaining (mostly compilation) issues after initial rebase of cep-15-accord on cep-21-tcm

dcapwell committed:
Operations.migrateReadRequiredOperations fails due to concurrent access when TransactionStatement is prepared

patch by David Capwell; reviewed by Ariel Weisberg, Caleb Rackliffe for CASSANDRA-18337

dcapwell committed:
CEP-15 (Accord) Expected reply message with verb ACCORD_INFORM_OF_TXNID_RSP but got ACCORD_SIMPLE_RSP

patch by David Capwell; reviewed by Caleb Rackliffe for CASSANDRA-18375

dcapwell committed:
CEP-15 (Accord) Original and recover coordinators may hit a race condition with PreApply where reads and writes are interleaved, causing one of the coordinators to see the writes from the other

patch by David Capwell; reviewed by Ariel Weisberg for CASSANDRA-18422

belliottsmith committed:
(Accord only) Permit nodes to join a cluster without the full transaction history

patch by Benedict; reviewed by Blake Eggleston for CASSANDRA-18523

dcapwell committed:
CEP-15: (C*) Accord message processing should avoid being passed on to a Stage and run directly in the messageing handler

patch by David Capwell; reviewed by Ariel Weisberg, Benedict Elliott Smith for CASSANDRA-18364

aweisberg committed:
Accord/non-Accord interoperability

Patch by Ariel Weisberg and Blake Eggleston; Reviewed by Blake Eggleston for CASSANDRA-18129

Co-authored-by: Blake Eggleston <beggleston@apple.com>

dcapwell committed:
CEP-15: (C*) Enhance in-memory FileSystem to work with mmap and support tests to add custom logic

patch by David Capwell; reviewed by Caleb Rackliffe for CASSANDRA-18485

dcapwell committed:
CEP-15: (C*) Add notion of CommandsForRanges and make this durable in C*

patch by David Capwell; reviewed by Ariel Weisberg for CASSANDRA-18519

Commit:
minor cleanup in Verb and BTreeSet

Commit:
removing unnecessary calls to ServerTestUtils.daemonInitialization() in a handful of tests

Commit:
make sure workspace.xml specifies a storagedir

iamaleksey committed:
CEP-15: Extend Accord MessageType with a side effect flag

patch by Aleksey Yeschenko; reviewed by Benedic Elliott Smith for
CASSANDRA-18561

iamaleksey committed:
CEP-15: Convert AccordStateCache cache from write-through to write-back

patch by Aleksey Yeschenko; reviewed by Blake Eggleston for
CASSANDRA-18563

bdeggleston committed:
CEP-15: Accord Bootstrap Integration

Patch by Blake Eggleston and Benedict Elliott Smith; Reviewed by David
Capwell for CASSANDRA-17101

CEP-15: Accord TCM integration

Patch by Blake Eggleston; Reviewed by David Capwell for CASSANDRA-18444

iamaleksey committed:
CEP-15: Minimize transaction state kept in system tables

patch by Aleksey Yeschenko; reviewed by Ariel Weisberg for CASSANDRA-18573

jonmeredith committed:
Several simulator fixes not yet merged to cep-15-accord

s c4bf1533b4 Add more JVM arg checks to the paxos simulation runner
s 1b650a0636 break circular dependency between FileSystemOwnershipCheck and CassandraRelevantProperties
s e0111748dd fix verify with-rng and interrupts
s f8475f2a18 make SimulatedFailureDetector deterministic
s ee77734bbe Deterministic transformation
s 890f23696a Add mechanism forsimulator to intercept password salting
s 5205b99874 Specify Paxos V2 because it is the only linearizable option

Patch by Ariel Weisberg; Reviewed by David Capwell for CASSANDRA-19008

belliottsmith committed:
Support state eviction (i.e. permit the state machine to erase transactions that are known to be applied across the cluster)

patch by Benedict Elliott Smith; reviewed by Ariel Weisberg, Aleksey Yeschenko, and David Capwell for CASSANDRA-18883

Co-authored-by: Benedict Elliott Smith <benedict@apache.org>
Co-authored-by: Ariel Weisberg <aweisberg@apple.com>
Co-authored-by: Aleksey Yeschenko <aleksey@apache.org>
Co-authored-by: David Capwell <dcapwell@gmail.com>

dcapwell committed:
CEP-15 (C*) When a host replacement happens don't loose the peer mapping right away (#3575)

patch by David Capwell; reviewed by Blake Eggleston for CASSANDRA-18764

dcapwell committed:
Fix Accord compaction purger tombstone logic

Accord compaction purgers see random slices of Accord state during compaction (based on randomly selected compaction inputs).

For at least the `durability` column in the `commands` table the tombstone being created when truncating was deleting the latest value since we can get enough information to truncate without actuall yhaving the latest `durability` value.

To fix we can wait to emit a tombstone until we are erasing the entire command row when truncating or truncating with outcome and meanwhile we can drop the extra columns that are no longer needed instead of using a tombstone. We don't need to emit cell tombstones we can drop them from the purger when processing each row.

patch by Ariel Weisberg; reviewed by David Capwell for CASSANDRA-18795

jacek-lewandowski committed:
CASSANDRA-18774: Fix pre-commit hook

aweisberg committed:
Allow exceptions to be propagated remotely

https://github.com/apache/cassandra-accord/pull/56

Patch by Ariel Weisberg; Reviewed by David Capwell for CASSANDRA-18779

dcapwell committed:
CEP-15 (C*): when loading commands that have empty waiting_on, make sure not to loose the partial deps (#3590)

patch by David Capwell; reviewed by Aleksey Yeschenko for CASSANDRA-18783

iamaleksey committed:
Fix CompactionAccordIteratorsTest

dcapwell committed:
CEP-15: (C*) NPE on restart as epoch isnt known but verb handler doesnt delay mapping to Node.Id

dcapwell committed:
CEP-15: (C*) accord.messages.BeginRecovery.RecoverNack#supersededBy is nullable but C* serializer doesn't expect null

Commit:
- Avoid persisting fragments that do not require completion as Updates in TxnWrite, as they can simply be pulled from PartialTxn when needed in Write#apply()
- Avoid serializing full TxnData instances to Accord state tables

patch by Caleb Rackliffe; reviewed by David Capwell, Benedict Elliot Smith, and Ariel Weisberg for CASSANDRA-18355

maedhroz committed:
ninja-fix: checkstyle fixes in VirtualTableTest, HistoryValidatorTest, SeedDefiner, RunStartDefiner, and Config

iamaleksey committed:
Fix Propagate serializer

maedhroz committed:
ninja-fix: minor post-TCM-rebase cleanup

maedhroz committed:
ninja-fix: exclude accord_demo.txt from rat checks

maedhroz committed:
Fix AccordJournalTest and ConfigCompatibilityTest, and revert the sendMessageWithFailureCB() callback to printing just the repair failure reason in its error message

dcapwell committed:
Get simulator working (again)

Co-authored-by: Ariel Weisberg <aweisberg@apple.com>
Co-authored-by: Benedict Elliott Smith <benedict@apache.org>

jacek-lewandowski committed:
CEP-15: Add Accord configuration stub

Patch by Jacek Lewandowski; reviewed by David Capwell for CASSANDRA-18221

belliottsmith committed:
Improve validation and address various discovered faults

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-19045

jacek-lewandowski committed:
CEP-15: Accord metrics

Patch by Jacek Lewandowski, reviewed by Caleb Rackliffe, David Capwell and Henrik Ingo for CASSANDRA-18580

dcapwell committed:
CEP-15: (C*) Implement TopologySorter to prioritise hosts based on DynamicSnitch and/or topology layout

patch by Blake Eggleston, David Capwell; reviewed by Blake Eggleston for CASSANDRA-18929

ifesdjeen committed:
Use pinned Harry version

bdeggleston committed:
Reduce command deps

Patch by Blake Eggleston; Reviewed by Benedict Elliott Smith for CASSANDRA-18784

aweisberg committed:
Ninja fix SlowMessageFuzzTest and InvertedIndexSearcherTest

Patch by Ariel Weisberg; Reviewed by David Capwell for CASANDRA-19023

aweisberg committed:
Fix Paxos V2 prepare response serialization

Patch by Ariel Weisberg; Reviewed by David Capwell for CASSANDRA-19023

iamaleksey committed:
Pre-requisite changes for CASSANDRA-18888

patch by Aleksey Yeschenko; reviewed by Ariel Weisberg for
CASSANDRA-18888

iamaleksey committed:
Quick fix for AccordCommandStoreTest#commandLoadSave()

dcapwell committed:
Ninja for CASSANDRA-19045: make sure to use https rather than git@ for submodule so its portable

dcapwell committed:
Ninja for CASSANDRA-19045: use the latest sha from trunk rather than an old one from 10 months ago

aweisberg committed:
Fix additional live migration/interop merge issues

s fc83325fa9 Fix AccordObjectSizes empty sizes
s 44bce6a08c Fix error handling when there are no column families to repair
s f1459540e5 Fix broken nowInSec deserialization
s 91c1befd1c Fix Mutation serializedSize
s b11467c200 Using simulated clock instead of raw nanoTime
s 6412b35924 Accord repair needs to use sentinel tokens
s d19c01ff01 Enable Accord repair with HappyPathFuzzTest
s 2da87f91b8 Implement abort in AccordRepairJob
s e12877c4fa move barrier to accord spec

Patch by Ariel Weisberg; Reviewed by David Capwell for CASSANDRA-19023

aweisberg committed:
Fix AccordCommandsPurger universal durability check

aweisberg committed:
Fix AccordJournal.Type interop entries

Patch by Ariel Weisberg; Reviewed by David Capwell for CASSANDRA-19024

bdeggleston committed:
Fix empty row check in command/tfk mutation methods

bdeggleston committed:
Fix CompactionAccordIteratorsTest

aweisberg committed:
Fix Mutation serializer

aweisberg committed:
Fix 5.1 rebase bugs

aweisberg committed:
Add back enforceInvariants=false to shouldCleanup in AccordCommandsPurger

aweisberg committed:
Fix StorageAttachedIndexDDLTest failing due to background Accord compactions

aweisberg committed:
Instance.receiveMessage should use sync

We mistakenly set this to async thinking async delivery was needed, but whether it is async or not is based on the boolean paramter passed to it.

Right now SchemaTest relies on it running sync, deserializaing in the other class loader and generating an error, and then handling that error from inside C*. This isn't how it actually works on failed deserialize (it's just a timeout), but the test relies on it so for now not changing it.

aweisberg committed:
Fix GuardrailTablesTest

aweisberg committed:
Don't use ImmediateExectuor in JVM dtests to process messages

For now just throwing it on MISC stage which is an unbounded thread pool that shouldn't deadlock

aweisberg committed:
Fix TombstoneCountWarningTest

bdeggleston committed:
Schema based accord fast path configuration

Patch by Blake Eggleston; Reviewed by David Capwell and Alex Petrov for CASSANDRA-19009

dcapwell committed:
(Accord): Bug fixes from CASSANDRA-18675 to better support adding keyspaces

patch by David Capwell; reviewed by Benedict Elliott Smith, Blake Eggleston for CASSANDRA-18804

dcapwell committed:
(Accord) NPE while trying to serialize FoundKnownMap as value is null half the time but unexpected while serializing

patch by David Capwell; reviewed by Blake Eggleston for CASSANDRA-19253

ifesdjeen committed:
Quick fix for AccordAddTableTest: make rejections catch up

bdeggleston committed:
fix fast path coordinator test

dcapwell committed:
Ninja: Added check that command.route is not null when SPL is trying to inform, saw this event in CI and need more details to know what happened

bdeggleston committed:
retry on Invalidated

bdeggleston committed:
fix ReducingRangeMapSerializer

bdeggleston committed:
improve accord error logging

bdeggleston committed:
don't make noise when fast path reconfigs are rejected

bdeggleston committed:
apply cfk update to both cfk caches

bdeggleston committed:
don't mark initialized cfk instances as updated

bdeggleston committed:
fix fast path update deadlock

belliottsmith committed:
CASSANDRA-18365: Protocol fixes

iamaleksey committed:
Fix LocalPartitioner duplication in *_for_key Accord system tables

patch by Aleksey Yeschenko; reviewed by Marcus Eriksson for
CASSANDRA-19265

bdeggleston committed:
load both key command histories for operations specifying keyhistory ALL

bdeggleston committed:
fix fast path test epoch wait

belliottsmith committed:
Fix SerializerSupport.reconstruct, and introduce RangeFactory

belliottsmith committed:
Fast single-partition "Ephemeral Reads"

Introduce a special kind of non-durable read that provides only per-key linearizable isolation; i.e. strict-serializable isolation for single partition-key reads.
This read creates only a happens-before edge, by collecting dependencies for execution and ensuring that execution happens strictly after these dependencies
have executed, but at no precise time otherwise. So later writes may be witnessed, and if multiple keys are read they may represents different points in time.

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-19305

belliottsmith committed:
Fix (de)serialization of WaitingOn into cache

belliottsmith committed:
Fix simulator logging and clean shutdown

belliottsmith committed:
Refactor CommandsForKey for efficiency, and to support transitive dependency elision

patch by Benedict; reviewed by Aleksey Yeshchenko for CASSANDRA-19310

bdeggleston committed:
CEP-15 (C*) Integrate accord with repair

Patch by Blake Eggleston; Reviewed by Ariel Weisberg and David Capwell for CASSANDRA-19472

aweisberg committed:
Add TODO for TxnQuery using wrong cluster metadata

belliottsmith committed:
perf improvements

bdeggleston committed:
Bootstrap/TCM integration test fixes

bdeggleston committed:
CEP-15: (C*) per-table transactional configuration

Patch by Blake Eggleston; Reviewed by Ariel Wesberg for CASSANDRA-19016

bdeggleston committed:
rebase fixes

bdeggleston committed:
fix MultiElementType pack/unpack accessor api

bdeggleston committed:
post-trunk rebase fixes

dcapwell committed:
Ninja for CASSANDRA-19305: Disable EphemeralRead by default to get benchmarks stable

dcapwell committed:
Accord: PreLoadContext must properly and consistently support ranges

patch by David Capwell; reviewed by Benedict Elliott Smith for CASSANDRA-19355

bdeggleston committed:
fix ReadCommand.serializedSize

bdeggleston committed:
revert async message delivery

dcapwell committed:
(Accord) Cassandra bootstrap no longer using the range txn and instead uses the sync point empty txn for reads

patch by David Capwell; reviewed by Blake Eggleston for CASSANDRA-19503

dcapwell committed:
AccordGens.rangeDeps did not enforce unique ranges, which caused tests to fail

bdeggleston committed:
fix read repair tests

dcapwell committed:
When jvm-dtest is shutting down an instance TCM retries block the shutdown causing the test to fail

patch by David Capwell; reviewed by Blake Eggleston for CASSANDRA-19514

maedhroz committed:
Prohibit counter column access in Accord transactions

patch by Caleb Rackliffe; reviewed by David Capwell for CASSANDRA-18987

dcapwell committed:
Accord: NPE in RangeDeps.forEach

patch by Benedict Elliott Smith, David Capwell; reviewed by Benedict Elliott Smith for CASSANDRA-19605

dcapwell committed:
Accord: Need to simulate Cassandra Journal in Accord BurnTest to detect issues earlier before they are seen in Cassandra

patch by Benedict Elliott Smith, David Capwell; reviewed by Benedict Elliott Smith, David Capwell for CASSANDRA-19618

maedhroz committed:
post-rebase fixes, mostly around CASSANDRA-19341 and CASSANDRA-19567

maedhroz committed:
make sure StreamSummarySerializer uses the appropriate IPartitioner during deserialization

maedhroz committed:
changing accord submodule branch back to trunk

iamaleksey committed:
Move preaccept expiration logic away from Agent

patch by Aleksey Yeschenko; reviewed by Alex Petrov and Benedict Elliott Smith for CASSANDRA-18888

maedhroz committed:
Baseline Diagnostic vtables for Accord

patch by Caleb Rackliffe; reviewed by David Capwell and Ariel Weisberg for CASSANDRA-18732

aweisberg committed:
Accord barrier/inclusive sync point fixes

Patch by Ariel Weisberg, Benedict Elliott Smith; reviewed by Benedict Elliott Smith for CASSANDRA-19641

dcapwell committed:
IndexOutOfBoundsException while serializing CommandsForKey

patch by David Capwell; reviewed by Blake Eggleston for CASSANDRA-19642

belliottsmith committed:
Introduce Periodic mode to Accord Journal

patch by Benedict; reviewed by Aleksey Yeschenko, Alex Petrov and David Capwell for CASSANDRA-19720

ifesdjeen committed:
Accord Journal Determinism: PreAccept replay stability

Patch by Alex Petrov; reviewed by Aleksey Yeschenko for CASSANDRA-19664

dcapwell committed:
CEP-15: (Accord) Bootstraps LocalOnly txn can not be recreated from SerializerSupport

patch by David Capwell; reviewed by Benedict Elliott Smith for CASSANDRA-19674

ifesdjeen committed:
Bring back Journal simulator (w/o Accord at least for now); add semaphore interceptor.

Patch by Alex Petrov; reviewed by David Capwell for CASSANDRA-19695.

aweisberg committed:
ApplyThenWaitUntilApplied supplies wrong epoch for executeAtEpoch

Patch by Ariel Weisberg; Reviewed by Benedict Elliott Smith for CASSANDRA-19687

bdeggleston committed:
ninja: fix cqlsh tests

ifesdjeen committed:
Accord Journal / Determinism

      * Store intermediate Command states in the log
      * Load Command states from the log
      * Simplify AccordJournal by removing framing

Patch by Alex Petrov; reviewed by Benedict Elliott Smith for CASSANDRA-19757

bdeggleston committed:
Don't run incremental repairs for consensus only repairs

Patch by Blake Eggleston; reviewed by David Capwell for CASSANDRA-19717

bdeggleston committed:
Move burn test read timestamp validation from replica to coordination

Patch by Blake Eggleston; reviewed by David Capwell for CASSANDRA-19288

dcapwell committed:
CEP-15: (Accord) SyncPoint timeouts become a Exhausted rather than a Timeout and doesn’t get retried

patch by David Capwell; reviewed by Ariel Weisberg for CASSANDRA-19718

dcapwell committed:
CEP-15: (Accord) When nodes are removed from a cluster, need to update topology tracking to avoid being blocked

patch by David Capwell; reviewed by Blake Eggleston for CASSANDRA-19719

bdeggleston committed:
ninja: don't fuzz test consensus only repairs

ifesdjeen commented on src/java/org/apache/cassandra/service/accord/AccordJournal.java:
This is an extremely early version. There's a bug in flusher: we do not seem to wait for last flush before we replay, which leads to some issues, so there's sleep here. But since this patch does not hook replay up to any prod code, I thought it would be fine to leave this in.

ifesdjeen commented on src/java/org/apache/cassandra/service/accord/SavedCommand.java:
I suggest we leave the discussed improvements, such as relying on status to know what to diff to a follow-up patch.

belliottsmith commented on src/java/org/apache/cassandra/journal/Journal.java:
Can't this introduce blocking, where a waiting thread is stuck behind a later write? Probably we should drain and sort the queue. It would be nice also to use something better than a `ConcurrentLinkedQueue` (but not critical)

belliottsmith commented on src/java/org/apache/cassandra/journal/Journal.java:
Can this not permit invoking `run` twice?

belliottsmith commented on src/java/org/apache/cassandra/journal/Journal.java:
It would be nice not to rely on additional atomics here. Is there any reason the journal thread won't promptly catch up any that could have run immediately?

belliottsmith commented on src/java/org/apache/cassandra/service/accord/AccordCommandStore.java:
if we return `null` the cache won't waste time handing to another thread for saving

ifesdjeen commented on src/java/org/apache/cassandra/journal/Journal.java:
It will catch up. Atomic's purpose was to avoid double-running, but it should never happen, just a precaution really.

ifesdjeen commented on src/java/org/apache/cassandra/service/accord/AccordCommandStore.java:
Great point!

belliottsmith commented on src/java/org/apache/cassandra/journal/Journal.java:
I mean if we don't invoke on the caller that submits - I agree this can permit double invocation, but we could just not do this?

ifesdjeen commented on src/java/org/apache/cassandra/journal/Journal.java:
Switched to MPSC queue, and added an array for draining. Will do a bigger improvement that will introduce callbacks native to Journal in a follow-up.

ifesdjeen commented on src/java/org/apache/cassandra/journal/Journal.java:
Yeah; the intention here was to make sure that we run the runnable only once. You are of course right that it journal thread will promptly catch up any that could have run immediately.

stonewhitener committed:
CEP-15 (Accord): When starting a transaction in a table where Accord is not enabled, should fail fast rather than fail with lack of ranges

patch by Youki Shiraishi; reviewed by Caleb Rackliffe, David Capwell for CASSANDRA-19759

ifesdjeen committed:
Fix remaining tests
